# Afame Technology: Human Resources Data Insights

In [1]:
# import required libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt

### The column names in the Titanic dataset are as follows:

    Age                         : The age of the employee.
    Attrition                   : Whether the employee has left the company or is still employed.
    BusinessTravel              : The frequency of business travel for the employee.
    DailyRate                   : The daily rate of pay for the employee.
    Department                  : The department in which the employee works.
    DistanceFromHome            : The distance of the employee's home from the workplace.
    Education                   : The level of education of the employee.
    EducationField              : The field of education the employee studied.
    EmployeeCount               : The number of employees in the company.
    EmployeeNumber              : A unique identifier for the employee.
    EnvironmentSatisfaction     : The satisfaction level of the employee with the work environment.
    Gender                      : The gender of the employee.
    HourlyRate                  : The hourly rate of pay for the employee.
    JobInvolvement              : The level of involvement the employee has in their job.
    JobLevel                    : The level of the employee's job within the company.
    JobRole                     : The specific role or position of the employee.
    JobSatisfaction             : The satisfaction level of the employee with their job.
    MaritalStatus               : The marital status of the employee.
    MonthlyIncome               : The monthly income of the employee.
    MonthlyRate                 : The monthly rate of pay for the employee.
    NumCompaniesWorked          : The number of companies the employee has worked for.
    Over18                      : Whether the employee is over 18 years old.
    OverTime                    : Whether the employee works overtime.
    PercentSalaryHike           : The percentage increase in salary for the employee.
    PerformanceRating           : The performance rating of the employee.
    RelationshipSatisfaction    : The satisfaction level of the employee with their relationships at work.
    StandardHours               : The standard number of working hours in a day.
    StockOptionLevel            : The level of stock options the employee has.
    TotalWorkingYears           : The total number of years the employee has worked.
    TrainingTimesLastYear       : The number of training sessions the employee attended last year.
    WorkLifeBalance             : The balance between work and personal life for the employee.
    YearsAtCompany              : The number of years the employee has been with the company.
    YearsInCurrentRole          : The number of years the employee has been in their current role.
    YearsSinceLastPromotion     : The number of years since the employee's last promotion.
    YearsWithCurrManager        : The number of years the employee has been with their current manager.

In [2]:
data = pd.read_csv("C:/Users/Surabhi/Downloads/HR Data.csv")
pd.set_option('display.max_columns', None)
print(data) 

      Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0       

In [3]:
data.shape

(1470, 35)

In [4]:
data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [6]:
# Loop through each column and print the value counts
for col in data.columns:
    print(f"Value counts for column '{col}':")
    print(data[col].value_counts())
    print("===================================")

Value counts for column 'Age':
Age
35    78
34    77
36    69
31    69
29    68
32    61
30    60
33    58
38    58
40    57
37    50
27    48
28    48
42    46
39    42
45    41
41    40
26    39
44    33
46    33
43    32
50    30
25    26
24    26
49    24
47    24
55    22
51    19
53    19
48    19
54    18
52    18
22    16
56    14
23    14
58    14
21    13
20    11
59    10
19     9
18     8
60     5
57     4
Name: count, dtype: int64
Value counts for column 'Attrition':
Attrition
No     1233
Yes     237
Name: count, dtype: int64
Value counts for column 'BusinessTravel':
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64
Value counts for column 'DailyRate':
DailyRate
691     6
408     5
530     5
1329    5
1082    5
       ..
650     1
279     1
316     1
314     1
628     1
Name: count, Length: 886, dtype: int64
Value counts for column 'Department':
Department
Research & Development    961
Sales              

In [7]:
data.drop(['DailyRate','EmployeeCount','EmployeeNumber','HourlyRate','Over18','StandardHours'], axis=1,inplace= True)

In [8]:
data.rename(columns ={'BusinessTravel':'Business_Travel','DistanceFromHome':'Distance_From_Home',
                             'EducationField':'Education_Field','EnvironmentSatisfaction':'Environment_Satisfaction',
                             'JobInvolvement':'Job_Involvement','JobLevel':'Job_Level',
                             'JobRole':'Job_Role','JobSatisfaction':'Job_Satisfaction',
                             'MaritalStatus':'Marital_Status','MonthlyIncome':'Monthly_Income',
                             'MonthlyRate':'Monthly_Rate','NumCompaniesWorked':'Num_Companies_Worked',
                             'OverTime':'Over_Time','PercentSalaryHike':'Percent_Salary_Hike',
                             'PerformanceRating':'Performance_Rating','RelationshipSatisfaction':'Relationship_Satisfaction',
                             'StockOptionLevel':'Stock_Option_Level','TotalWorkingYears':'Total_Working_Years',
                             'TrainingTimesLastYear':'Training_Times_Last_Year','WorkLifeBalance':'Work_Life_Balance',
                             'YearsAtCompany':'Years_At_Company','YearsInCurrentRole':'Years_In_Current_Role',
                             'YearsSinceLastPromotion':'Years_Since_Last_Promotion','YearsWithCurrManager':'Years_With_Curr_Manager'},inplace = True)

In [9]:
data.describe()

Age  Distance_From_Home    Education  Environment_Satisfaction  \
count  1470.000000         1470.000000  1470.000000               1470.000000   
mean     36.923810            9.192517     2.912925                  2.721769   
std       9.135373            8.106864     1.024165                  1.093082   
min      18.000000            1.000000     1.000000                  1.000000   
25%      30.000000            2.000000     2.000000                  2.000000   
50%      36.000000            7.000000     3.000000                  3.000000   
75%      43.000000           14.000000     4.000000                  4.000000   
max      60.000000           29.000000     5.000000                  4.000000   

       Job_Involvement    Job_Level  Job_Satisfaction  Monthly_Income  \
count      1470.000000  1470.000000       1470.000000     1470.000000   
mean          2.729932     2.063946          2.728571     6502.931293   
std           0.711561     1.106940          1.102846     4707.956783   
min           1.000000     1.000000          1.000000     1009.000000   
25%           2.000000     1.000000          2.000000     2911.000000   
50%           3.000000     2.000000          3.000000     4919.000000   
75%           3.000000     3.000000          4.000000     8379.000000   
max           4.000000     5.000000          4.000000    19999.000000   

       Monthly_Rate  Num_Companies_Worked  Percent_Salary_Hike  \
count   1470.000000           1470.000000          1470.000000   
mean   14313.103401              2.693197            15.209524   
std     7117.786044              2.498009             3.659938   
min     2094.000000              0.000000            11.000000   
25%     8047.000000              1.000000            12.000000   
50%    14235.500000              2.000000            14.000000   
75%    20461.500000              4.000000            18.000000   
max    26999.000000              9.000000            25.000000   

       Performance_Rating  Relationship_Satisfaction  Stock_Option_Level  \
count         1470.000000                1470.000000         1470.000000   
mean             3.153741                   2.712245            0.793878   
std              0.360824                   1.081209            0.852077   
min              3.000000                   1.000000            0.000000   
25%              3.000000                   2.000000            0.000000   
50%              3.000000                   3.000000            1.000000   
75%              3.000000                   4.000000            1.000000   
max              4.000000                   4.000000            3.000000   

       Total_Working_Years  Training_Times_Last_Year  Work_Life_Balance  \
count          1470.000000               1470.000000        1470.000000   
mean             11.279592                  2.799320           2.761224   
std               7.780782                  1.289271           0.706476   
min               0.000000                  0.000000           1.000000   
25%               6.000000                  2.000000           2.000000   
50%              10.000000                  3.000000           3.000000   
75%              15.000000                  3.000000           3.000000   
max              40.000000                  6.000000           4.000000   

       Years_At_Company  Years_In_Current_Role  Years_Since_Last_Promotion  \
count       1470.000000            1470.000000                 1470.000000   
mean           7.008163               4.229252                    2.187755   
std            6.126525               3.623137                    3.222430   
min            0.000000               0.000000                    0.000000   
25%            3.000000               2.000000                    0.000000   
50%            5.000000               3.000000                    1.000000   
75%            9.000000               7.000000                    3.000000   
max           40.000000              18.000000 

In [10]:
data.shape

(1470, 29)

In [11]:
data.duplicated().sum()

0

In [12]:
data.isnull().sum()

Age                           0
Attrition                     0
Business_Travel               0
Department                    0
Distance_From_Home            0
Education                     0
Education_Field               0
Environment_Satisfaction      0
Gender                        0
Job_Involvement               0
Job_Level                     0
Job_Role                      0
Job_Satisfaction              0
Marital_Status                0
Monthly_Income                0
Monthly_Rate                  0
Num_Companies_Worked          0
Over_Time                     0
Percent_Salary_Hike           0
Performance_Rating            0
Relationship_Satisfaction     0
Stock_Option_Level            0
Total_Working_Years           0
Training_Times_Last_Year      0
Work_Life_Balance             0
Years_At_Company              0
Years_In_Current_Role         0
Years_Since_Last_Promotion    0
Years_With_Curr_Manager       0
dtype: int64

### Encoding

In [13]:
## Converting categorical variables into numerical variables using label Encoding
from sklearn.preprocessing import LabelEncoder

cat_cols = ['Attrition', 'Business_Travel', 'Department', 'Education_Field', 'Gender','Job_Role','Marital_Status','Over_Time',]

le = LabelEncoder()

for col in cat_cols:
    data[col] = le.fit_transform(data[col])

In [14]:
data 

Age  Attrition  Business_Travel  Department  Distance_From_Home  \
0      41          1                2           2                   1   
1      49          0                1           1                   8   
2      37          1                2           1                   2   
3      33          0                1           1                   3   
4      27          0                2           1                   2   
...   ...        ...              ...         ...                 ...   
1465   36          0                1           1                  23   
1466   39          0                2           1                   6   
1467   27          0                2           1                   4   
1468   49          0                1           2                   2   
1469   34          0                2           1                   8   

      Education  Education_Field  Environment_Satisfaction  Gender  \
0             2                1                         2       0   
1             1                1                         3       1   
2             2                4                         4       1   
3             4                1                         4       0   
4             1                3                         1       1   
...         ...              ...                       ...     ...   
1465          2                3                         3       1   
1466          1                3                         4       1   
1467          3                1                         2       1   
1468          3                3                         4       1   
1469          3                3                         2       1   

      Job_Involvement  Job_Level  Job_Role  Job_Satisfaction  Marital_Status  \
0                   3          2         7                 4               2   
1                   2          2         6                 2               1   
2                   2          1         2                 3               2   
3                   3          1         6                 3               1   
4                   3          1         2                 2               1   
...               ...        ...       ...               ...             ...   
1465                4          2         2                 4               1   
1466                2          3         0                 1               1   
1467                4          2         4                 2               1   
1468                2          2         7                 2               1   
1469                4          2         2                 3               1   

      Monthly_Income  Monthly_Rate  Num_Companies_Worked  Over_Time  \
0               5993         19479                     8          1   
1               5130         24907                     1          0   
2               2090          2396                     6          1   
3               2909         23159                     1          1   
4               3468         16632                     9          0   
...              ...           ...                   ...        ...   
1465            2571         12290                     4          0   
1466            9991         21457                     4          0   
1467            6142          5174                     1          1   
1468            5390         13243                     2          0   
1469            4404         10228                     2          0   

      Percent_Salary_Hike  Performance_Rating  Relationship_Satisfaction  \
0                      11                   3                          1   
1                      23                   4                          4   
2                      15                   3                          2   
3                      11                   3                          3   
4                      12                   3                          4   
...               

Based on the code and analysis, we can draw the following conclusions about the human resources dataset:

1. The dataset contains information on 1470 employees, with 29 columns detailing various aspects like age, attrition, department, education, job role, performance metrics, income, and other work-related factors.

2. There are no duplicate rows or null values present in the dataset, indicating that the data is clean and ready for analysis.

3. The dataset contains both numerical and categorical variables. Numerical variables include age, distance from home, income, years of experience, and performance ratings, while categorical variables include attrition status, business travel, department, education field, gender, job role, and marital status.

4. The categorical variables have been converted into numerical form using label encoding, making the data suitable for machine learning algorithms that require numerical input.

5. The descriptive statistics provide insight into the distribution of numerical variables, such as the average age of employees being around 37 years, the average total working years being around 11 years, and the majority of employees having a performance rating of 3 (on a scale of 1 to 4).

6. The value counts for categorical variables reveal patterns like the majority of employees being in the Research & Development department, most employees having a bachelor's or master's degree, and a significant portion of the workforce being married.

Overall, the code has prepared the human resources dataset for further analysis, such as building predictive models for employee attrition, job satisfaction, or performance evaluation. The cleaned and transformed dataset can now be used for exploratory data analysis, feature engineering, and applying machine learning techniques to gain insights and make data-driven decisions related to human resource management.